### PR TITLE
Dinner button: remember about good food

### PR DESCRIPTION
**Pull request by MelokG000**
## About The Pull Request
image
Adds a one-use button on each map that makes an announcement that reminds crew that they should eat good food
## Why It's Good For The Game

Many people forget about chef's food and uses junkfood from vendors or raw fruits, a reminder button could help with that
## Changelog
:cl:
add: chef can now use the dinner button to remind crew about food (available on every station!)
/:cl:

### CODE REVIEW

**Comment by ArcaneMusic** [comment]
Howdy, you'll need to run the mapmerge script so that your map files are converted to our format (And so that you don't have 1.2 million lines changed!).

The instructions are on the repo under `tools\mapmerge2\README.md`, and in the future you can also use the mapmerge hook or mapping tools like StrongDMM to avoid having to do it. Should be an easy fix.

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	. = ..()
+
+/obj/machinery/foodbutton/attack_hand(mob/living/user)
+	. = ..()
+	announce()
+
+/obj/machinery/foodbutton/proc/announce(mob/living/user)
+	if(button_used)
+		return FALSE
+	minor_announce("Attention crew! It's lunchtime! Head over for a delicious meal prepared just for you. Bon appétit!", "Kitchen announcement")
```
Comment (Wallemations) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:22]: I thought this was a dinner button not a lunch button

**Comment by Wallemations** [comment]
Wait I just looked at the code

What's stopping other people from pressing this button besides the chef
This feels more like it should require an ID with chef access tbh
And it's available from roundstart too? So if you press it immediately there's just no way to actually announce this later?

As much as I think this is a really funny idea I don't think its implementation is very airtight. 
Maybe it could be an ID-locked button that can only be opened when a certain amount of time has passed or something, idfk

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	playsound(src, button_sound, 70, TRUE, -1)
+	button_used = TRUE
+	return TRUE
```
Comment (Wallemations) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:25]: Why are you returning TRUE or FALSE for a custom proc that's not used in an if statement

**Review by Wallemations — APPROVED** [review]
Sprite looks fine

Code review isn't my job but I at least skimmed it over and gave my 2 cents

**Comment by CCC23** [comment]
> image what the FUCK happened here

DINNER

**Comment by the-orange-cow** [comment]
Likely beyond the scope of this pull, but it would be interesting if this had some kind of interaction or shared the function of https://github.com/tgstation/tgstation/pull/85242.

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	icon_state = "button_supper"
+	layer = OBJ_LAYER
+	anchored = TRUE
+	pass_flags = PASSTABLE // Able to place on tables
+	var/button_used = FALSE // only one announcement per shift
+	var/button_sound = 'sound/machines/pda_button2.ogg'
+
+/obj/machinery/foodbutton/Initialize(mapload)
+	. = ..()
+
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:14]:
```suggestion
```

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+/obj/machinery/foodbutton/attack_hand(mob/living/user)
+	. = ..()
+	announce()
+
+/obj/machinery/foodbutton/proc/announce(mob/living/user)
+	if(button_used)
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:20]:
```suggestion
	if(!allowed(user))
		balloon_alert(user, "access denied!")
		return
	if(button_used)
		balloon_alert(user, "already used!")
```
It should be locked to those with proper kitchen access to prevent random asshats from untimely using it.

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	var/button_used = FALSE // only one announcement per shift
+	var/button_sound = 'sound/machines/pda_button2.ogg'
+
+/obj/machinery/foodbutton/Initialize(mapload)
+	. = ..()
+
+/obj/machinery/foodbutton/attack_hand(mob/living/user)
+	. = ..()
+	announce()
+
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:18]:
```suggestion
/obj/machinery/foodbutton/attack_hand(mob/living/user)
	. = ..()
	announce()

/obj/machinery/foodbutton/attack_paw(mob/living/user)
	. = ..()
	announce()

```
Monkeys...

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/obj/machinery/foodbutton
+	name = "dinner button"
+	desc = "Casts an announcement to the station, reminding the crew about food."
+	icon = 'icons/obj/service/kitchen.dmi'
+	icon_state = "button_supper"
+	layer = OBJ_LAYER
+	anchored = TRUE
+	pass_flags = PASSTABLE // Able to place on tables
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:8]:
```suggestion
	pass_flags = PASSTABLE // Able to place on tables
	req_access = list(ACCESS_KITCHEN)
```

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+/obj/machinery/foodbutton/Initialize(mapload)
+	. = ..()
+
+/obj/machinery/foodbutton/attack_hand(mob/living/user)
+	. = ..()
+	announce()
+
+/obj/machinery/foodbutton/proc/announce(mob/living/user)
+	if(button_used)
+		return FALSE
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:21]:
```suggestion
		return
	if(world.time - SSticker.round_start_time < STOP_SERVING_BREAKFAST)
		balloon_alert(user, "too early!")
		return
```
We shouldn't let chefs use the button roundstart; it's still breakfast time and they probably haven't cooked much yet.

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/obj/machinery/foodbutton
+	name = "dinner button"
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:2]:
```suggestion
	name = "lunch button"
```

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+/obj/machinery/foodbutton/attack_hand(mob/living/user)
+	. = ..()
+	announce()
+
+/obj/machinery/foodbutton/proc/announce(mob/living/user)
+	if(button_used)
+		return FALSE
+	minor_announce("Attention crew! It's lunchtime! Head over for a delicious meal prepared just for you. Bon appétit!", "Kitchen announcement")
+	playsound(src, button_sound, 70, TRUE, -1)
+	button_used = TRUE
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:24]: It should have a different appearance when used, to give visual information that it's been used (i.e. button pushed down, screen turned off)

File: code/modules/food_and_drinks/machinery/dinner_button.dm [hunk 10]
```diff
@@ -0,0 +1,25 @@
+/obj/machinery/foodbutton
+	name = "dinner button"
+	desc = "Casts an announcement to the station, reminding the crew about food."
+	icon = 'icons/obj/service/kitchen.dmi'
+	icon_state = "button_supper"
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/dinner_button.dm:5]: If you could rename the icon state to button_lunch, that'd be less confusing.

File: tgstation.dme
Comment (Ghommie) [inline tgstation.dme:1]: Did you properly fork the repository on your device? You're somehow pushing changes to the line endings of each line in the .dme (probably from LF to CR LF), which normally should be prevented by the git settings of this repository. Please check that out, thank you.

**Review by Ghommie — CHANGES_REQUESTED** [review]
If you don't know how to use the map-merger to prevent byond from switching back to the default crappy format, you need to go to the tools/mapmerge2 folder. From there you can read the [readme](https://github.com/tgstation/tgstation/blob/master/tools/mapmerge2/README.md) to know how to prevent these issues in the future (it's really frikkin' easy by the by, you just run a couple .bat to install the hooks and it'll do it for you). I also suggest you download [StrongDMM](https://github.com/SpaiR/StrongDMM/releases/tag/v2.12.0.alpha) instead of using Byond's built-in map editor, also because it doesn't deconvert the current map format.

**Comment by Ghommie** [comment]
Ok, it seems the situation is a bit more, umh, basilar than what I've assumed so far...

Normally, when contributing to tgstation, SS13 or about any github repository, you've to install git or a git client. This allows you to create commits (like, packets of code changes) and push them to a branch of your fork (copy) of the repository (or just repo), and then compare them with another branch (which doesn't need to be from your repo, it can be the main repo or another fork of that repo) to make a pull request.

What you did wrong however was not downloading and properly setting up a git client to use to push your changes. Instead you merely downloaded the repo, extracted the folder, done your edits and then took advantage of github features that let you upload your changes and perform webedits. Which is what is giving you some troubles with the map files (.dmm) as well as the environment file (.dme). Your commits are all verified, which is usually a symptom that you've done so called webedits, that is unless you've a GPG key set, which is unlikely if you are not a professional contributing to some big-shot open source projects like chromium,

Now, you don't have to be good at coding to contribute to tgstation. Byond code is made to easy, and therefore also quite stupid. However, git is still pretty much fundamental to avoid problems like this. Also visual studio code is good, but you can still do most without it. It'll just be more painful.

TL;DR: [make sure to check this hackmd guide](https://hackmd.io/@tgstation/HJ8OdjNBc).

**Comment by Ghommie** [comment]
This is a tad older and outdated, but I'm running out of guides that can help you (I know there's another from some forum of BeeStation but I'm not sure I can find it): https://tgstation13.org/wiki/Setting_up_git

I know it isn't easy, but please try to do some troubleshooting on why git isn't working for you. See if you're having problem with the client or perhaps need one. Using git alone is kinda bonkers if you aren't accostumed to using a command line interface, so seek something with a friendlier user interface.
